---
title: "Integrative sparse Partial Least Squares Discriminant Analysis multiomics modeling"
output: 
  html_document:
    df_print: paged
    toc: true
    toc_depth: 2
    toc_float: true
    number_sections: true
author: Michael Chimenti
date: "`r format(Sys.time(), '%d %B %Y')`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Introduction 

**This is an attempt to use the RNA-seq and Proteome data to create a predictive model of COVID severity.**  We will use the sPLS method as before, but with an additional "DA" (discriminant analysis) term. We are interested in evaluating correlations and relationships between the RNA-seq and proteomic data in our 4 covid cohorts (COVID/ICU, COVID/Inpatient, COVID/Outpatient, No-COVID,Inpatient).  We are looking at the two-way integration before attempting the 3-way modeling incorporating the exome data. PLS was designed with a canonical (exploratory) approach and a regression (explanatory) approach in mind. Partial Least Squares – Discriminant Analysis (PLS-DA) was hence developed to allow the powerful PLS algorithm to be used for classification [1, 2]. It performs very similarly to PLS, just that the response vector y contains categorical vectors rather than continuous vectors. PLS-DA has the same advantages that PLS does, such that it operates efficiently over large dataframes and is not negatively influenced by collinearity.

# Methods 

We are following the general pattern in this vignette (http://mixomics.org/methods/spls-da/) and this case study (http://mixomics.org/case-studies/splsda-srbct-case-study/).    

# Analysis 

## Imports

```{r libraries, warning=FALSE, message=FALSE}
library(tidyverse)
library(here)
library(mixOmics)

here::i_am("analysis/mixomics_sPLS_DA_predictive.Rmd")
```

## Load data 

```{r}

rna <- readr::read_csv(here("data","munged_data","rnaseq_samples_in_rows_normalized.csv"))
prot <- readr::read_csv(here("data","munged_data","proteomics_PLAS_ONLY_samples_in_rows_norm_NAMEFIX.csv"))
meta <- readr::read_csv(here("data","munged_data","metadata_masterkey.csv"))
```

## Wrangle metadata

Make the sample names match.  Not all samples specified in the metadata are found in the datafiles.  

```{r}
new_name <- rna$name %>% str_split(., pattern="-", simplify = TRUE) %>% paste0("RL_SARS_",.)
new_name <- new_name[1:length(rna$name)]
rna$sample_ID <- new_name 

prot <- prot %>% rename(sample_ID=Assay)

### There are 76 of 168 total proteomic datasets (we are excluding saliva datasets from consideration which is the reason for the mismatch)
sum(prot$sample_ID %in% meta$sample_ID)

###  There are 97 of 97 RNAseq files specified in meta 
                        #number of RNA files in meta  == #samples in rna found in meta 
assertthat::assert_that(sum(meta$sample_type == "RNA") == sum(rna$sample_ID %in% meta$sample_ID))


### Each data table needs a "master key" to determine sample of origin, without regards to data type 
master_key_dict <- meta %>% dplyr::select(c(sample_ID, master_key))
prot2 <- left_join(prot, master_key_dict, by="sample_ID") %>% drop_na()
rna2 <- left_join(rna, master_key_dict, by="sample_ID")

### We are missing some data...need to confirm with Ryan lab 
intersect(prot2$master_key, rna2$master_key)
setdiff(prot2$master_key, rna2$master_key)

```


### Convert data tables to named matrices 

```{r}
match <- intersect(prot2$master_key, rna2$master_key)
prot3 <- prot2[prot2$master_key %in% match,] %>% dplyr::select(-c(sample_ID)) %>% as.data.frame()
rownames(prot3) <- prot3$master_key
prot3 <- prot3 %>% dplyr::select(-master_key)

rna3 <- rna2[rna2$master_key %in% match,] %>% dplyr::select(-c(name, sample_ID)) %>% as.data.frame()
rownames(rna3) <- rna3$master_key
rna3 <- rna3 %>% dplyr::select(-master_key)

## check dimensions match 
dim(rna3)
dim(prot3)

## The metadata categories are the same whether RNA or Plasma 
meta_3 <- meta_3 <- meta[meta$master_key %in% match & meta$sample_type=="RNA",]

## Sanity check 
rownames(rna3) == meta_3$master_key
rownames(prot3) == meta_3$master_key
```

### Initial sPLS-DA model 

For starters, we'll include the COVID NEG samples just to see how the model looks: 

```{r}
spls.da.rna.covid <- splsda(X=rna3, Y=meta_3$cohort_status, ncomp=10)
spls.da.pro.covid <- splsda(X=prot3, Y=meta_3$cohort_status, ncomp=10)

```

```{r}
# plot the samples projected onto the first two components of the PLS-DA subspace from RNA-seq data
plotIndiv(spls.da.rna.covid , comp = 1:2, 
          group = meta_3$cohort_status, ind.names = FALSE,  # colour points by class
          ellipse = TRUE, # include 95% confidence ellipse for each class
          legend = TRUE, title = 'RNA-seq PLSDA with confidence ellipses')


plotIndiv(spls.da.pro.covid , comp = 1:2, 
          group = meta_3$cohort_status, ind.names = FALSE,  # colour points by class
          ellipse = TRUE, # include 95% confidence ellipse for each class
          legend = TRUE, title = 'Proteome PLSDA with confidence ellipses')

# use the max.dist measure to form decision boundaries between classes based on PLS-DA data
background = background.predict(spls.da.rna.covid, comp.predicted=2, dist = "max.dist")

# plot the samples projected onto the first two components of the PLS-DA subspace
plotIndiv(spls.da.rna.covid, comp = 1:2, rep.space = "XY-variate",
          group = meta_3$cohort_status, ind.names = FALSE, # colour points by class
          background = background, # include prediction background for each class
          legend = TRUE, title = " (b) PLSDA with prediction background")
```

### sPLS-DA without the COVID NEG samples 

```{r}
mask <- meta_3$covid_status != "NEG"
rna4 <- rna3[mask,]
prot4 <- prot3[mask,]
meta_4 <- meta_3[mask,]

## check dimensions match 
dim(rna4)
dim(prot4)

## Sanity check 
rownames(rna4) == meta_4$master_key
rownames(prot4) == meta_4$master_key
```

### Single omic sPLS-DA models 

Previous tuning of the sPLS models suggests that 2 components and 15-20 features will be appropriate: 
Number of PLS Components = 2
Features to keep = 15 

```{r}
spls.da.rna.covid <- splsda(X=rna4, Y=meta_4$cohort_status, ncomp=2, keepX = 15)
spls.da.pro.covid <- splsda(X=prot4, Y=meta_4$cohort_status, ncomp=2, keepX = 15)
```

### Visualizing simple omics sPLS-DA models 

#### RNAseq alone 

```{r}
# plot the samples projected onto the first two components of the PLS-DA subspace from RNA-seq data
plotIndiv(spls.da.rna.covid , comp = 1:2, rep.space = "XY-variate",
          group = meta_4$cohort_status, ind.names = FALSE,  # colour points by class
          ellipse = TRUE, # include 95% confidence ellipse for each class
          legend = TRUE, title = 'RNA-seq sPLS-DA projection in XY space')


```

```{r}
# use the max.dist measure to form decision boundaries between classes based on PLS-DA data
background_rna = background.predict(spls.da.rna.covid, comp.predicted=2, dist = "max.dist")

# plot the samples projected onto the first two components of the PLS-DA subspace
plotIndiv(spls.da.rna.covid, comp = 1:2, rep.space = "XY-variate",
          group = meta_4$cohort_status, ind.names = FALSE, # colour points by class
          background = background_rna, # include prediction background for each class
          legend = TRUE, title = " RNA-seq sPLS-DA with decision boundary")

```


In this plot, we can see the samples in the PLS reduction to first 2 components, 
looking at the mean of XY coordinates together.  This is similar to PCA, but because of the 
Discriminant Analysis (DA), we have a decision boundary space that predicts sample types.  

#### Proteomics alone 

```{r}

plotIndiv(spls.da.pro.covid , comp = 1:2, rep.space = "XY-variate",
          group = meta_4$cohort_status, ind.names = FALSE,  # colour points by class
          ellipse = TRUE, # include 95% confidence ellipse for each class
          legend = TRUE, title = 'Proteome sPLS-DA projection in XY space')

```

```{r}
background_pro = background.predict(spls.da.pro.covid, comp.predicted=2, dist = "max.dist")
plotIndiv(spls.da.pro.covid, comp = 1:2, rep.space = "XY-variate",
          group = meta_4$cohort_status, ind.names = FALSE, # colour points by class
          background = background_pro, # include prediction background for each class
          legend = TRUE, title = " Proteome sPLS-DA with decision boundary")

```

NB: "s22_day1" is an outlier for the COVID POS Inpatient cohort in Proteomics, "s24_day1" is an outlier for RNA-seq.  

### Important Features in single 'omics models 

#### RNAseq 

***When undergoing most dimension reduction methods in the mixOmics package, latent components are produced. These latent components are defined by their corresponding loading vectors, which are vectors with the weight of each original variable's contribution to the corresponding latent component. Greater absolute values in this loading vector means that a given variable has a greater “importance”.The plotLoadings() function allows for the visualisation of this importance using a bar plot, where the most relevant original variables (those with the greatest absolute loading value) will be at the bottom of the plot. It can generate a few different types of plot depending on the context. If the sparse variant of a method is used, only the selected variables will be shown.***

These are the genes that most contribute to the XY variate plots and the decision boundary above: 

```{r}
plotLoadings(spls.da.rna.covid, comp = 1, ndisplay = 15, title = "Genes selected by sparse PLS for comp 1")
```


```{r, fig.height=9, fig.width=9}
plotVar(spls.da.rna.covid, cutoff = 0.65, cex = c(2.5), title = "Genes highly correlated with component 1 and 2")
```

```{r, fig.height=9, fig.width=9}

cim(spls.da.rna.covid, comp=1 )
```
#### Proteomics 

These are the proteins that most contribute to the XY variate plots and the decision boundary above: 

```{r}
plotLoadings(spls.da.pro.covid, comp = 1, ndisplay = 15, title = "Proteins selected by sparse PLS for comp 1")
```
```{r, fig.height=9, fig.width=9}

cim(spls.da.pro.covid, comp=1 )
```

### Multiblock sPLS-DA 

Now that we've performed classification procedure with PLS for both RNA-seq and Proteomics data, let's combine both datasets into an N-integration multiblock model for our classification task and see if that improves performance further.  

```{r}
X <- list(rna = rna4, prot = prot4)
Y <- meta_4$cohort_status

list.keepX = list(rna = c(20,20), prot = c(20, 20)) 

des <- matrix(data = c(0,0.1,0.1,0), nrow=2, ncol=2)
diablo.covid <- block.splsda(X=X, Y=Y, ncomp=5, keepX=list.keepX, design = des)
```

```{r}
# plot the contributions of each feature to each dimension
plotLoadings(diablo.covid, ncomp = 1) 

plotIndiv(diablo.covid, rep.space = "XY-variate", 
          group = meta_4$cohort_status, ind.names = FALSE, 
          ellipse = TRUE,
          legend = TRUE, title = 'N-Int sPLS-DA with confidence ellipses')


```

```{r}
# run component number tuning with repeated CV
perf.diablo = perf(diablo.covid, validation = 'Mfold', 
                   folds = 5, nrepeat = 20) 

plot(perf.diablo) # plot output of tuning

# set the optimal ncomp value
ncomp = perf.diablo$choice.ncomp$WeightedVote["Overall.BER", "centroids.dist"] 
# show the optimal choice for ncomp for each dist metric
perf.diablo$choice.ncomp$WeightedVote 
```

```{r}
# set grid of values for each component to test
test.keepX = list (rna = c(5:9, seq(10, 18, 2), seq(20,30,5)), 
                   prot = c(5:9, seq(10, 18, 2), seq(20,30,5)))

ncomp = 2

# run the feature selection tuning
tune.diablo = tune.block.splsda(X = X, Y = Y, ncomp = ncomp, 
                              test.keepX = test.keepX, design = des,
                              validation = 'Mfold', folds = 5, nrepeat = 1,
                              dist = "centroids.dist")

list.keepX = tune.diablo$choice.keepX # set the optimal values of features to retain
list.keepX
```

```{r}
final.diablo.covid = block.splsda(X=X, Y=Y, ncomp=2, keepX=list.keepX,design=des)
```


```{r}
plotDiablo(final.diablo.covid, ncomp=1)
plotDiablo(final.diablo.covid, ncomp=2)
```

```{r}
plotIndiv(final.diablo.covid, ind.names = FALSE, legend=TRUE, title="Diablo Block PLots")
```

```{r}
plotArrow(final.diablo.covid, legend=TRUE, title="DIABLO arrow plot")
```
***In the arrow plot above, the start of the arrow indicates the centroid between all data sets for a given sample and the tips of the arrows indicate the location of that sample in each block. Such graphics highlight the agreement between all data sets at the sample level.***

Outpatient covid has better agreement between centroid of model components and individual datasets.  There is a subset of Covid Pos Inpatient that also has higher agreement, and several samples have lower agreement (s24_day3, s22_day1, s23_day1, etc...)

```{r, fig.width=9,fig.height=9,dpi=300}
circosPlot(final.diablo.covid, cutoff = 0.55, line = TRUE,
           color.blocks= c('darkorchid', 'lightgreen'), 
           size.labels = 2, size.variables = 0.75)
```

```{r, fig.height=11,fig.width=11,dpi=200}
cimDiablo(final.diablo.covid, margins = c(5,5))
```

### Testing performance 

```{r}
# run repeated CV performance evaluation
perf.diablo = perf(final.diablo.covid, validation = 'Mfold', progressBar = TRUE,
                   folds = 5, nrepeat = 25, ncpus=4) 

perf.diablo$MajorityVote.error.rate
perf.diablo$WeightedVote.error.rate
```



```{r}
auc.splsda = auroc(final.diablo.covid, roc.block = "prot", 
                   roc.comp = 2, print = FALSE)

auc.splsda = auroc(final.diablo.covid, roc.block = "rna", 
                   roc.comp = 2, print = FALSE)

```

